src/Utils/openAuthModal.ts

```diff
@@ -0,0 +1,192 @@
+import * as Schema from "Artsy/Analytics/Schema"
+import { Mediator } from "Artsy/SystemContext"
+import { stringify } from "qs"
+import { data as sd } from "sharify"
+
+interface AuthReason {
+  entity: {
+    id: string
+    name: string
+  }
+  contextModule: Schema.ContextModule
+  intent: AuthModalIntent
+}
+
+export enum AuthModalIntent {
+  FollowArtist = "FollowArtist",
+  FollowPartner = "FollowPartner",
+  SaveArtwork = "SaveArtwork",
+}
+
+export const openAuthModal = (mediator: Mediator, reason: AuthReason) => {
+  let handled = false
+
+  if (sd.IS_MOBILE) {
+    const intent = getMobileAuthIntent(reason)
+    if (intent) {
+      openMobileAuth(intent)
+      handled = true
+    }
+  } else if (mediator) {
+    const intent = getDesktopAuthIntent(reason)
+    if (intent) {
+      openDesktopAuth(mediator, intent)
+      handled = true
+    }
+  }
+
+  if (!handled) {
+    window.location.href = "/login"
+  }
+}
+
+function openMobileAuth(intent) {
+  const params = stringify(intent)
+  const href = `/sign_up?redirect-to=${window.location}&${params}`
+
+  window.location.href = href
+}
+
+function openDesktopAuth(mediator, intent) {
+  mediator.trigger("open:auth", intent)
+}
+
+interface MobileIntent {
+  action: string
+  contextModule: Schema.ContextModule | string
+  intent: string
+  kind: string
+  objectId: string
+  signUpIntent: string
+  trigger: string
+  entityName: string
+}
+
+function getMobileAuthIntent(reason: AuthReason): MobileIntent {
+  switch (reason.intent) {
+    case AuthModalIntent.FollowArtist:
+      return getMobileIntentToFollowArtist(reason)
+    case AuthModalIntent.FollowPartner:
+      return getMobileIntentToFollowPartner(reason)
+    case AuthModalIntent.SaveArtwork:
+      return getMobileIntentToSaveArtwork(reason)
+    default:
+      return undefined
+  }
+}
+
+function getMobileIntentToFollowArtist({
+  contextModule,
+  entity,
+}: AuthReason): MobileIntent {
+  return {
+    action: "follow",
+    contextModule,
+    intent: "follow artist",
+    kind: "artist",
+    objectId: entity.id,
+    signUpIntent: "follow artist",
+    trigger: "click",
+    entityName: entity.name,
+  }
+}
+
+function getMobileIntentToFollowPartner({
+  contextModule,
+  entity,
+}: AuthReason): MobileIntent {
+  return {
+    action: "follow",
+    contextModule,
+    intent: "follow partner",
```

Comment:
This `intent` (and the `signUpIntent` on line 104) [_used_ to be "follow gallery"](https://github.com/artsy/reaction/pull/2303/files#diff-070e554f536ad96b2b31ad0e4544b513L68), not "follow partner". 

I changed it to "follow partner", because [this code in Force](https://github.com/artsy/force/blob/74fb00704a7c33b0a7b02c5bf328159990b65a9b/src/desktop/apps/auth2/routes.ts#L62-L63) indicates that it is expecting "follow partner". 

A side-effect of this change is that the mobile auth screen now says "Sign in to follow partners" instead of "Sign in to Artsy" when trying to follow a partner.